beluga_amcl/test/test_amcl_node.cpp

```diff
@@ -0,0 +1,861 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <tf2_ros/create_timer_ros.h>
+
+#include <beluga_amcl/private/amcl_node.hpp>
+
+#include <lifecycle_msgs/msg/state.hpp>
+
+namespace
+{
+
+using namespace std::chrono_literals;
+
+/// Spin a group of nodes until a condition is met.
+/**
+ * \param predicate The stop condition.
+ * \param timeout Maximum time to spin.
+ * \param nodes The nodes to spin.
+ * \return True if the condition was met. False if it timed out.
+ */
+template<class Predicate, class Rep, class Period, class ... Nodes>
+bool spin_until(
+  Predicate && predicate,
+  const std::chrono::duration<Rep, Period> & timeout,
+  const std::shared_ptr<Nodes> & ... nodes)
+{
+  rclcpp::executors::SingleThreadedExecutor executor;
+  (executor.add_node(nodes->get_node_base_interface()), ...);
+
+  auto start = std::chrono::high_resolution_clock::now();
+  while (std::chrono::high_resolution_clock::now() - start < timeout) {
+    executor.spin_some();
+    if (predicate()) {
+      return true;
+    }
+  }
+  return false;  // timed out
+}
+
+/// Spin a group of nodes until a condition is met with a default timeout.
+/**
+ * \param predicate The stop condition.
+ * \param nodes The nodes to spin.
+ * \return True if the condition was met. False if it timed out.
+ */
+template<class Predicate, class ... Nodes>
+bool spin_until(
+  Predicate && predicate,
+  const std::shared_ptr<Nodes> & ... nodes)
+{
+  return spin_until(std::forward<Predicate>(predicate), 10s, nodes ...);
+}
+
+/// Spin a group of nodes for a given duration of time.
+/**
+ * \param predicate The stop condition.
+ * \param duration Time to spin.
+ */
+template<class Rep, class Period, class ... Nodes>
+void spin_for(
+  const std::chrono::duration<Rep, Period> & duration,
+  const std::shared_ptr<Nodes> & ... nodes)
+{
+  spin_until([]() {return false;}, duration, nodes ...);
```

Comment:
@nahueespinosa nit^N!:
```suggestion
  const auto duration_is_over = []() {return false;};
  spin_until(duration_is_over, duration, nodes ...);
```

Reply:
Done in https://github.com/Ekumen-OS/beluga/pull/194/commits/00b9631edb630940b1e87c2e57e0919da9320a67!